FEATURE: Added recording on mousedown

burstMembrane opened the pull request:
This PR adds recording and upload to firebase on mousedown/touchdown as referenced in feature request #5 

It records for as long as user holds down their finger on the screen/holds mouse and is directly attached to the animation handler, should provide a fairly seamless experience.

burstMembrane committed:
added mouse recording experiment

burstMembrane committed:
added recording on mousedown with hard recorder limit

burstMembrane committed:
tweaked recording permission check

wezzahd commented:
I just tested this branch and there are a couple of issues with the sound capture:

* I like the pitch mappings.  Could we make the pitches based on screen Y-axis so low notes at bottom and high at top?  I think that will give users a clue to the location of where their pixel will appear on the outdoor screen. 

* How does the recording max limit work?  Is that based on time between mouse holds?  Or max number of recording per session?  

* Mousehold does not trigger permission request.  Need to click the mic icon in the menu to trigger.  Can we get permissions during loading similar to cam?  (https://github.com/bunjilplace2021/realness/issues/5#issuecomment-761989341).

* Mousehold sound capture not working with chrome?  Chrome log attached:

''
TypeError: Cannot read property 'decodeAudioData' of undefined
    at soundmain.js:14392
    at :3000/async http:/127.0.0.1:3000/sound/soundmain.js:13299
soundmain.js:14486 Uncaught (in promise) TypeError: Cannot read property 'sampleRate' of undefined
    at soundmain.js:14486
    at new Promise (<anonymous>)
    at resampleBuffer (soundmain.js:14478)
    at soundmain.js:13303
''

wezzahd commented:
Also could we print to console when sound is captured via mousedown?  (similar to when mic icon is clicked).

burstMembrane committed:
logging to console

burstMembrane committed:
fixed merge conflicts

burstMembrane committed:
added variable length recording

burstMembrane committed:
added additional recorder checks

burstMembrane commented:
@wezzahd  Issues with mousedown recording should should be fixed in latest deploy-preview. Also added logging to the console when user recordings are started.

[https://deploy-preview-14--adoring-mayer-3b5b15.netlify.app/](url)

I added a record limit of 3 so that we don't overload the database with user recordings. So after three user recordings have been made and played back, it reverts to the normal playback algorithm. This might help with scaling later on??


I'm looking into mapping sounds low to high as well. will merge a commit when that's working :)

burstMembrane committed:
added minimum recording time

burstMembrane committed:
mapped sound to inverse pixel height

burstMembrane committed:
removed mobile check for masterbus chorus

wezzahd commented:
> @wezzahd Issues with mousedown recording should should be fixed in latest deploy-preview. Also added logging to the console when user recordings are started.
> 
> [https://deploy-preview-14--adoring-mayer-3b5b15.netlify.app/](url)
> 
> I added a record limit of 3 so that we don't overload the database with user recordings. So after three user recordings have been made and played back, it reverts to the normal playback algorithm. This might help with scaling later on??
> 
> I'm looking into mapping sounds low to high as well. will merge a commit when that's working :)

@burstMembrane Probably a good idea to turn off the mousehold animation if limit has been reached and no sound is being recorded.  Is there a boolean variable for the max threshold that I can use in the particle system?

wezzahd committed:
merge dev

wezzahd commented:
@burstMembrane I've noticed on my macbook that some mouseclicks are registering as mousedown so the recording gets triggered when i click. 
The particle animation for the mousedown gets triggered if the mouse button is held longer than approx one second.  Could we do something similar for mic capture?

burstMembrane commented:
@wezzahd I have a recordingLimitReached Boolean that switches to true. I'll add it to the window object so that you can query it in your code for example 
```
If(!window.recordingLimitReached)
{
// Do mouse hold animation
}
```

Also I noticed there is a mouse count that increments on mouse down. I'll change the code to fire the event listener to synchronise with the mouse hold animation. 

If possible we could make the recording icon an absolute element in the top left that is usually hidden and fades in when the user is making a recording? Might mess with the design a little but I might have a crack at it in a new commit and we can see if it works.

wezzahd commented:
> @wezzahd I have a recordingLimitReached Boolean that switches to true. I'll add it to the window object so that you can query it in your code for example
> 
> ```
> If(!window.recordingLimitReached)
> {
> // Do mouse hold animation
> }
> ```
> 
> Also I noticed there is a mouse count that increments on mouse down. I'll change the code to fire the event listener to synchronise with the mouse hold animation.
> 
> If possible we could make the recording icon an absolute element in the top left that is usually hidden and fades in when the user is making a recording? Might mess with the design a little but I might have a crack at it in a new commit and we can see if it works.

@burstMembrane Maybe hold off on the recording indicator element.  I'd like to have some kind of animation in the particle system that indicates to the user that the recording is being captured when the mousehold event is triggered.  At the moment the particle size and alpha oscillates on mouse hold but I need to come up with something better.

burstMembrane committed:
Merge branch 'dev' of github.com:bunjilplace2021/realness into mouserecording

burstMembrane committed:
tweaked recording permission check

burstMembrane committed:
added additional checks

burstMembrane committed:
added timout to mousepress

burstMembrane committed:
removed fsevents dependency

burstMembrane committed:
desktop safari permissions check

burstMembrane committed:
changed safari permissions check

burstMembrane committed:
changed safari permissions check

burstMembrane committed:
changed safari permissions check

wezzahd commented:
> @wezzahd I have a recordingLimitReached Boolean that switches to true. I'll add it to the window object so that you can query it in your code for example
> 
> ```
> If(!window.recordingLimitReached)
> {
> // Do mouse hold animation
> }
> ```
> 
> Also I noticed there is a mouse count that increments on mouse down. I'll change the code to fire the event listener to synchronise with the mouse hold animation.
> 
> If possible we could make the recording icon an absolute element in the top left that is usually hidden and fades in when the user is making a recording? Might mess with the design a little but I might have a crack at it in a new commit and we can see if it works.

@burstMembrane I've come up with a new mouse hold animation in a seperate branch.  Is there a callback or window object I can use to trigger the animation when the mic recording and also completes?

burstMembrane committed:
fixed merge conflicts

burstMembrane committed:
fixed merge conflicts

burstMembrane committed:
merged webcamcheck branch

burstMembrane committed:
Merge branch 'dev' into mouserecording

burstMembrane commented:
> > @wezzahd I have a recordingLimitReached Boolean that switches to true. I'll add it to the window object so that you can query it in your code for example
> > ```
> > If(!window.recordingLimitReached)
> > {
> > // Do mouse hold animation
> > }
> > ```
> > 
> > 
> > Also I noticed there is a mouse count that increments on mouse down. I'll change the code to fire the event listener to synchronise with the mouse hold animation.
> > If possible we could make the recording icon an absolute element in the top left that is usually hidden and fades in when the user is making a recording? Might mess with the design a little but I might have a crack at it in a new commit and we can see if it works.
> 
> @burstMembrane I've come up with a new mouse hold animation in a seperate branch. Is there a callback or window object I can use to trigger the animation when the mic recording and also completes?

@wezzahd there's probably only one check you need to do which is to check if the window.recording boolean has changed. In my latest commit this occurs only when the user has allowed mouse recording and the recording is happening so it would work to sync the animation. It also won't switch to true after the recording limit is reached or if the sound is muted. 

example pseudo code:

```
if(window.recording)
{
// do mouse hold animation
}


```

wezzahd commented:
@burstMembrane Thanks for that.  Has the window.recording boolean been committed yet?  I can only see the window.recordingLimitReached boolean.  I'll commit the mouse recording branch to dev once that's added.

wezzahd commented:
@burstMembrane Hey Liam,  there seems to be a bug with the sound using Safari (both iOS & desktop).  Unable to unmute audio (just spinning icon).  Sometimes the sound loads but there is a lot of distortion and crackle.  I can't see any errors in the console log.

wezzahd commented:
@burstMembrane The only error coming up is 
``
[Error] Failed to load resource: the server responded with a status of 404 () (version.js.map, line 0)  https://deploy-preview-14--adoring-mayer-3b5b15.netlify.app.
``

burstMembrane commented:
@wezzahd I'll take a look today. Was sound not working in safari desktop or iPhone?

burstMembrane commented:
> @burstMembrane The only error coming up is
> `[Error] Failed to load resource: the server responded with a status of 404 () (version.js.map, line 0) https://deploy-preview-14--adoring-mayer-3b5b15.netlify.app.`

@wezzahd was working on safari on my Mac. I'll take a look today might be a permissions thing. Should be ready to merge on next commit!

wezzahd commented:
> > @burstMembrane The only error coming up is
> > `[Error] Failed to load resource: the server responded with a status of 404 () (version.js.map, line 0) https://deploy-preview-14--adoring-mayer-3b5b15.netlify.app.`
> 
> @wezzahd was working on safari on my Mac. I'll take a look today might be a permissions thing. Should be ready to merge on next commit!

Seems to be a weird intermittent bug.  Sound loaded OK when I opened the site in a private window in Firefox, Chrome and Safari.

burstMembrane committed:
added additional fallbacks for iphone and safari users

burstMembrane committed:
added additional fallbacks for iphone and safari users

burstMembrane committed:
added additional fallbacks for iphone and safari users

burstMembrane committed:
removed references to redundant recordbutton

burstMembrane committed:
added playbutton & merged didactic panel

burstMembrane committed:
added playbutton & merged didactic panel

wezzahd committed:
Merge branch 'dev' into mouserecording

wezzahd committed:
Revert "Merge branch 'dev' into mouserecording"

This reverts commit 1d28d2d54c53621206f42bc042e70b04a87d40cd, reversing
changes made to 9fa7e5506e68847f9bb69334f3da8e1fdd6b79d0.
